### PR TITLE
NEW Allow manipulating eagerloading queries

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

Adds a backwards-compatible way to pre-filter, pre-sort, and apply basically any other type of transformation you want to eagerloaded data.

Note that I have opted to keep the existing public API, since the alternative suggested in the issue would not be usable in templates. I tried consulting with other devs on slack per the acceptance criteria but there's really not anyone who's familiar with eagerloading yet.

Applying a limit in the callback is explicitly disallowed because it wouldn't be applied per relation list, it'd be applied to the whole eager query for reach relation.

## Manual testing steps

There's benchmarking code in the linked issue.

## Issues

- https://github.com/silverstripe/silverstripe-framework/issues/10929

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/ORM/DataList.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 use SilverStripe\ORM\DataQuery;
 use SilverStripe\ORM\ArrayList;
 use SilverStripe\ORM\Filters\SearchFilterable;
+use Symfony\Component\Validator\Tests\Constraints\Limit;
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]:
```suggestion
```

File: src/ORM/DataList.php [hunk 2]
```diff
@@ -1480,8 +1545,11 @@ public function eagerLoad(...$relationChains): static
                 // Keep track of what we've seen before so we don't accidentally add a level 1 relation
                 // (e.g. "Players") to the chains list when we already have it as part of a longer chain
                 // (e.g. "Players.Teams")
-                $list->eagerLoadAllRelations[$item] = $item;
+                $list->eagerLoadAllRelations[$item] ??= null;
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1563]: This change does two things:
1. Default the value to `null`, so we can do a simply nullable check to see if there was a callback set for that chain or not
2. Doesn't override existing values - if a callback was explicitly set, we don't want to remove it here.

File: src/ORM/DataList.php [hunk 3]
```diff
@@ -1480,8 +1545,11 @@ public function eagerLoad(...$relationChains): static
...
             }
+            // Set the callback for this chain
+            $list->eagerLoadAllRelations[$relationChain] = $callback;
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1566]: Note that this *will* override a callback with null if someone adds the same chain twice, the second time without a callback. For example:

```php
$list = $list->eagerLoad(['MyRelation.AnotherRelation' => $someCallback]);
$list = $list->eagerLoad('MyRelation.AnotherRelation');
```

This allows complex branching logic, where some branches can remove the filters/sorting/etc that other branches might apply.

File: src/ORM/DataList.php [hunk 4]
```diff
@@ -1308,13 +1317,19 @@ private function fetchEagerLoadManyMany(
...
+        $fetchedRows = $fetchList->getFinalisedQuery();
 
         foreach ($fetchedRows as $row) {
             $fetchedRowsArray[$row['ID']] = $row;
             $fetchedIDs[] = $row['ID'];
         }
 
+        // Get the join records so we can correctly identify which children belong to which parents
+        // This also holds extra fields data
+        $joinRows = DB::query('SELECT * FROM "' . $joinTable . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ') AND ' . $childIDField . ' IN (' . implode(',', $fetchedIDs) . ')');
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]: Note we're now only getting the join records that are applicable to the potentially-filtered-down set of relations.

File: src/ORM/DataList.php [hunk 5]
```diff
@@ -1115,6 +1116,11 @@ private function fetchEagerLoadHasOne(
         string $relationName,
         string $relationType
     ): array {
+        // Throw exception if developers try to filter/sort a has_one relation
+        if ($this->eagerLoadAllRelations[$relationChain] !== null) {
+            throw new LogicException("Cannot filter/sort $relationType relation $relationName");
+        }
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]: Say something like "transform" or "manipulate" instead of "filter/sort", since developers could be trying to do just about anything in the callback.

File: src/ORM/DataList.php [hunk 6]
```diff
@@ -1115,6 +1115,11 @@ private function fetchEagerLoadHasOne(
         string $relationName,
         string $relationType
     ): array {
+        // Throw exception if developers try to manipulate a has_one relation as a list
+        if ($this->eagerLoadAllRelations[$relationChain] !== null) {
+            throw new LogicException("Cannot manipulate eagerloading query for $relationType relation $relationName");
+        }
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1121]: You can't go `$myRecord->MyHasOne()->filter()`, for example. This would be the eagerloading equivalent of that if we allowed it.

File: src/ORM/DataList.php [hunk 7]
```diff
@@ -1278,10 +1290,6 @@ private function fetchEagerLoadManyMany(
         $fetchedIDs = [];
         $eagerLoadedLists = [];
 
-        // Get the join records so we can correctly identify which children belong to which parents
-        // This also holds extra fields data
-        $joinRows = DB::query('SELECT * FROM "' . $joinTable . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ')');
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1283]: Moved to be after fetching the records we care about, and is now filtered and sorted based on that.

File: src/ORM/DataList.php [hunk 8]
```diff
@@ -1308,13 +1316,28 @@ private function fetchEagerLoadManyMany(
...
+        // This also holds extra fields data
+        $fetchedIDsAsString = implode(',', $fetchedIDs);
+        $joinRows = DB::query(
+            'SELECT * FROM "' . $joinTable
+            // Only get joins relevant for the parent list
+            . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ')'
+            // Exclude any children that got filtered out
+            . ' AND ' . $childIDField . ' IN (' . $fetchedIDsAsString . ')'
+            // Respect sort order of fetched items
+            . ' ORDER BY FIELD(' . $childIDField . ', ' . $fetchedIDsAsString . ')'
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1338]: The sorting is important here because the join rows are what we're looping through - so the order of these join rows determines the order of the records in the resultant `EagerLoadedList`

File: src/ORM/DataList.php [hunk 9]
```diff
@@ -1398,6 +1421,33 @@ private function addEagerLoadedDataToParent(
...
 
+    /**
+     * NOTE: Do not change `DataList` to `static` in this method signature.
+     * Subclasses of DataList must still accept DataList arguments and return DataList!
+     */
+    private function manipulateEagerLoadingQuery(
+        DataList $fetchList,
+        string $relationChain,
+        string $relationType
+    ): DataList {
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1432]: Discussed with Steve - he prefers `DataList` here over `self`.
I'm agnostic - I think the comment needs to be there for our future selves either way, and both IDEs and static analysis will ultimately treat them the same.

File: src/ORM/DataList.php [hunk 10]
```diff
@@ -1447,8 +1497,17 @@ private function fillEmptyEagerLoadedRelations(
      * You can specify nested relations by using dot notation, and you can also pass in multiple relations.
      * When specifying nested relations there is a maximum of 3 levels of relations allowed i.e. 2 dots
      *
-     * Example:
+     * Examples:
+     * <code>
      * $myDataList->eagerLoad('MyRelation.NestedRelation.EvenMoreNestedRelation', 'DifferentRelation')
+     * </code>
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1503]: Adding `<code>` tags here is an unrelated improvement for API docs since I'm adding another example here.

File: src/ORM/DataList.php [hunk 11]
```diff
@@ -1458,8 +1517,29 @@ private function fillEmptyEagerLoadedRelations(
     public function eagerLoad(...$relationChains): static
     {
         $list = clone $this;
-        foreach ($relationChains as $relationChain) {
-            // Don't add any relations we've added before
+
+        // If an array is passed in directly, treat it as though $relationChains wasn't spread.
+        if (count($relationChains) === 1 && is_array($relationChains[array_key_first($relationChains)])) {
+            $relationChains = $relationChains[array_key_first($relationChains)];
+        }
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1524]: Handles the difference between passing strings (original API, useful for templates) and passing an array (enhanced API, necessary if applying manipulations)

File: src/ORM/DataList.php [hunk 12]
```diff
@@ -1458,8 +1517,29 @@ private function fillEmptyEagerLoadedRelations(
...
+                $callback = null;
+            }
+
+            // Reject non-callable in associative array
+            if ($callback !== null && !is_callable($callback)) {
+                throw new LogicException(
+                    'Value of associative array must be a callable.'
+                    . 'If you don\'t want to pre-filter the list, use an indexed array.'
+                );
+            }
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1539]: Bit of added type safety - it would cause an error later on, but throwing out a clear message about what's wrong here will improve the debugging experience

File: src/ORM/DataQuery.php [hunk 13]
```diff
@@ -842,6 +842,14 @@ public function limit(?int $limit, int $offset = 0): static
         return $this;
     }
 
+    /**
+     * Get the current limit of this query.
+     */
+    public function getLimit(): array
+    {
+        return $this->query->getLimit();
+    }
```
Comment (GuySartorelli) [inline src/ORM/DataQuery.php]: This new API is necessary for us to explicitly disallow calling `limit()` in an eagerloading manipulation, since at best allowing that would provide unexpected results.

File: tests/php/ORM/DataListEagerLoadingTest.php [hunk 14]
```diff
@@ -897,7 +898,7 @@ public function testEagerLoadingFilteredList(string $iden, array $eagerLoad): vo
     {
         $this->createEagerLoadData(5);
         $filter = ['Title:GreaterThan' => 'obj 0'];
-        $dataList = EagerLoadObject::get()->filter($filter)->eagerLoad(...$eagerLoad);
+        $dataList = EagerLoadObject::get()->filter($filter)->eagerLoad($eagerLoad);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataListEagerLoadingTest.php:901]: I've swapped a few existing tests to pass the array directly, so we have coverage for non-associative arrays. I've left others alone to retain coverage for passing strings.

File: tests/php/ORM/DataListEagerLoadingTest.php [hunk 15]
```diff
@@ -1288,5 +1289,226 @@ public function testEagerLoadingSharedRelations()
                 $this->assertInstanceOf(EagerLoadedList::class, $eagerLoaded2);
             }
         }
+        $this->assertGreaterThan(1, $loop1Count);
+        $this->assertGreaterThan(1, $loop2Count);
```
Comment (GuySartorelli) [inline tests/php/ORM/DataListEagerLoadingTest.php:1293]: Unrelated change - the test was meant to have these originally which is why those variables were there.

File: tests/php/ORM/DataListEagerLoadingTest.php [hunk 16]
```diff
@@ -1288,5 +1289,226 @@ public function testEagerLoadingSharedRelations()
...
+    {
+        $this->expectException(LogicException::class);
+        $this->expectExceptionMessage(
+            'Value of associative array must be a callable.'
+            . 'If you don\'t want to pre-filter the list, use an indexed array.'
+        );
+        EagerLoadObject::get()->eagerLoad(['HasManyEagerLoadObjects' => 'HasManyEagerLoadObjects']);
+    }
+
+    public function provideNoLimitEagerloadingQuery(): array
```
Comment (emteknetnz) [inline tests/php/ORM/DataListEagerLoadingTest.php]: Do we need to test/implement `belongs_many_many`?

`many_many_through`?

Comment (emteknetnz) [inline tests/php/ORM/DataListEagerLoadingTest.php]:
```suggestion
    public function provideNoLimitEagerLoadingQuery(): array
```

File: tests/php/ORM/DataListEagerLoadingTest.php [hunk 17]
```diff
@@ -1288,5 +1289,226 @@ public function testEagerLoadingSharedRelations()
...
+            ],
+        ];
+    }
+
+    /**
+     * Tests that attempting to limit an eagerloading query will throw an exception.
+     *
+     * @dataProvider provideNoLimitEagerloadingQuery
+     */
+    public function testNoLimitEagerloadingQuery(string $relation, string $relationType, callable $callback): void
```
Comment (emteknetnz) [inline tests/php/ORM/DataListEagerLoadingTest.php]:
```suggestion
     * @dataProvider provideNoLimitEagerLoadingQuery
     */
    public function testNoLimitEagerLoadingQuery(string $relation, string $relationType, callable $callback): void
```

File: src/ORM/DataQuery.php [hunk 18]
```diff
@@ -842,6 +842,14 @@ public function limit(?int $limit, int $offset = 0): static
         return $this;
     }
 
+    /**
+     * Get the current limit of this query.
+     */
+    public function getLimit(): array
```
Comment (emteknetnz) [inline src/ORM/DataQuery.php]: This method seems really out of place in the class - there are no other "get*" methods doing the equivalent  getting of something on $this->query.

It also seems like a strange way to enfoce NOT doing something somwhere else

Is there another way to achieve the same thing somewhere e.g. in EagerLoadedList?

File: tests/php/ORM/DataQueryTest.php [hunk 19]
```diff
@@ -879,4 +879,38 @@ public function testWithUsingDataQueryAppliesRelations()
         // This will throw an exception if it fails - it passes if there's no exception.
         $dataQuery->execute();
     }
+
+    public function provideGetLimit(): array
```
Comment (emteknetnz) [inline tests/php/ORM/DataQueryTest.php]: Ideally we'd get rid of this test - see my comment on DataQuery

**Review by emteknetnz — CHANGES_REQUESTED** [review]

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I got the following error when testing this:

```
// WITH EAGERLOADING, FILTER/SORT IN DB
// 0.1115, 0.1077, 0.1019, 0.1110 -- 0.1080s (avg) (~29% faster than without eagerloading and ~85% faster than filtering in PHP)
foreach (MyDataObject::get()->eagerLoad(['MySubDataObjects' => fn (DataList $list) => $list->filter('Title:PartialMatch', 1)->Sort('Title', 'DESC')]) as $do) {
    $do->MySubDataObjects()->toArray();
}
```

```
ERROR [Emergency]: Uncaught TypeError: {closure}(): Argument #1 ($list) must be of type DataList, SilverStripe\ORM\DataList given, called in /var/www/vendor/silverstripe/framework/src/ORM/DataList.php on line 1435
IN GET dev/build
```

**Comment by GuySartorelli** [comment]
> I got the following error when testing this:

The code I copied into the issue is missing a `use` statement - I must have been overly aggressive when tidying it up.

File: src/ORM/DataQuery.php (hunk 18 above)
Comment (GuySartorelli) [reply src/ORM/DataQuery.php]: > It also seems like a strange way to enfoce NOT doing something somwhere else

It's obviously not _directly_ enforcing that, but we do need to get the current `limit` value in order to check if it has been set or not.

> Is there another way to achieve the same thing somewhere e.g. in EagerLoadedList?

I assume you mean in `DataList` - by the time you hit `EagerLoadedList` all of the data has been loaded already, so it's way too late.
Also, no, without reflection there isn't another way to do it that I'm aware of.

File: tests/php/ORM/DataListEagerLoadingTest.php (hunk 16 above)
Comment (GuySartorelli) [reply tests/php/ORM/DataListEagerLoadingTest.php]: No harm in it, will do.

Comment (GuySartorelli) [reply tests/php/ORM/DataListEagerLoadingTest.php]: Done

File: tests/php/ORM/DataListEagerLoadingTest.php (hunk 17 above)
Comment (GuySartorelli) [reply tests/php/ORM/DataListEagerLoadingTest.php]: Done

File: tests/php/ORM/DataListEagerLoadingTest.php (hunk 16 above)
Comment (GuySartorelli) [reply tests/php/ORM/DataListEagerLoadingTest.php]: Done - all of the many_many relation types are built the same so was pretty easy to refactor for the extra types.

File: src/ORM/DataQuery.php (hunk 18 above)
Comment (emteknetnz) [reply src/ORM/DataQuery.php]: Remove, this new API since we can avoid adding it, just use `$limit = $fetchList->dataQuery->query()->getLimit();` instead

File: src/ORM/DataList.php [hunk 20]
```diff
@@ -1398,6 +1421,33 @@ private function addEagerLoadedDataToParent(
...
+        $filterCallback = $this->eagerLoadAllRelations[$relationChain];
+        if ($filterCallback !== null) {
+            $fetchList = $filterCallback($fetchList);
+        }
+        if (!($fetchList instanceof DataList)) {
+            throw new LogicException(
+                "Eagerloading callback for $relationType relation $relationChain must return a DataList."
+            );
+        }
+        $limit = $fetchList->dataQuery->getLimit();
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        $limit = $fetchList->dataQuery->query()->getLimit();
```
That's used in first() and last()

It will trigger DataQuery::getFinalisedQuery(), though the performance loss from that will be minuscule

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/ORM/DataList.php (hunk 20 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: I really don't like this (it feels untidy), but given it's already used in this class for similar reasons it's not worth fighting about.

Done

File: tests/php/ORM/DataQueryTest.php (hunk 19 above)
Comment (GuySartorelli) [reply tests/php/ORM/DataQueryTest.php]: Done

File: src/ORM/DataQuery.php (hunk 18 above)
Comment (GuySartorelli) [reply src/ORM/DataQuery.php]: Feels untidy but https://github.com/silverstripe/silverstripe-framework/pull/11140#discussion_r1495066358

**Review by emteknetnz — APPROVED** [review]
